src/link.rs

```diff
@@ -141,41 +142,77 @@ where
             LinkInner::C(_) | LinkInner::Ca(_, _) | LinkInner::Ica(_, _, _) => {
                 return Ok(LinkCompound(borrow))
             }
-            LinkInner::Ia(_, _) => {
-                #[cfg(feature = "persistence")]
-                {
-                    // re-borrow mutable
-                    drop(borrow);
-                    let mut borrow = self.inner.borrow_mut();
-                    if let LinkInner::Ia(id, anno) =
-                        mem::replace(&mut *borrow, LinkInner::Placeholder)
-                    {
-                        match Persistence::get(&id) {
-                            Ok(generic) => {
-                                let compound = C::from_generic(&generic)?;
-                                *borrow =
-                                    LinkInner::Ica(id, Rc::new(compound), anno);
-
-                                // re-borrow immutable
-                                drop(borrow);
-                                let borrow = self.inner.borrow();
-
-                                Ok(LinkCompound(borrow))
-                            }
-                            Err(PersistError::Canon(e)) => Err(e),
-                            Err(
-                                PersistError::Io(_) | PersistError::Other(_),
-                            ) => {
-                                // TODO: log errors to the backend
-                                Err(CanonError::NotFound)
+            LinkInner::Ia(id, _) => {
```

Comment:
The whole branch that is following contains a lot of if / match nested, some code duplication, and it's not very readable. Could you try to do a bit of refactoring here? In some cases I think you can take advantages of guards for example. reducing the nesting.
Also, a bit more of documentation (comments) would be nice.

Reply:
I still believe there is too much nesting, but the comments help. We can land as is and trying to improve later.